File: daily_problems/2025/11/1127/personal_submission/cf106197c_Koishii.py
import sys
input = lambda: sys.stdin.readline().rstrip()

def query(x, y):
    print("?", x, y, flush = True)
    return int(input())

def soviet():
    n = int(input())
    pos1 = 1
    for i in range(2, n + 1):
        if query(pos1, i) != pos1:
            pos1 = i

    cur = []
    ans = []

    for i in range(1, n + 1):
        if i != pos1:
            cur.append(i)

    while len(cur):
        ncur = []
        prime_val = cur[0]
        for i in range(1, len(cur)):
            val = query(prime_val, cur[i])

            if val == pos1:
                ncur.append(cur[i])
            else:
                prime_val = val

        cur = ncur
        ans.append(prime_val)

    print("!", len(ans), end = ' ')
    print(*ans)

soviet()

File: daily_problems/2025/11/1127/cf106197c_Koishii.cpp
#include <bits/stdc++.h>
using namespace std;

int query(int x, int y) {
	std::cout << "? " << x << ' ' << y << std::endl;
	int res; std::cin >> res;
	return res;
}

void soviet() {
	int n; std::cin >> n;
	int pos1 = 1;

	for (int i = 2; i <= n; i++) {
		if (query(pos1, i) != pos1) {
			pos1 = i;
		}
	}
	std::vector<int> cur, ans;
	for (int i = 1; i <= n; i++) {
		if (i != pos1) {
			cur.emplace_back(i);
		}
	}
	while (cur.size()) {
		std::vector<int> ncur;
		int prime_val = cur[0];
		for (int i = 1; i < cur.size(); i++) {
			int val = query(prime_val, cur[i]);

			if (val == pos1) ncur.emplace_back(cur[i]);
			else prime_val = val;
		}

		cur = ncur;
		ans.emplace_back(prime_val);
	}

	std::cout << "! " << ans.size();
	for (int &x: ans) {
		std::cout << ' ' << x;
	}
}

int main() {
	int M_T = 1; //std::ios::sync_with_stdio(false); std::cin.tie(nullptr);
//	std::cin >> M_T;
	while (M_T--) { soviet(); }
	return 0;
}
